README.md
swaggerator
===

Installation
---
You can install the package via composer:

```bash
composer require peterzaccha/swaggerator
```

If you are using Laravel in a version < 5.5, the service provider must be registered as a next step:

```php
// config/app.php
'providers' => [
    ...
    Peterzaccha\Swaggerator\SwaggeratorServiceProvider::class
];
```
You can publish L5-Swagger's config and view files into your project by running:

```bash
$ php artisan vendor:publish --provider "L5Swagger\L5SwaggerServiceProvider"
```

For Laravel >=5.5, no need to manually add `L5SwaggerServiceProvider` into config. It uses package auto discovery feature. Skip this if you are on >=5.5, if not:

Open your `AppServiceProvider` (located in `app/Providers`) and add this line in `register` function
```php
$this->app->register(\L5Swagger\L5SwaggerServiceProvider::class);
```
or open your `config/app.php` and add this line in `providers` section
```php
L5Swagger\L5SwaggerServiceProvider::class,
```

in .env

```dotenv
L5_SWAGGER_GENERATE_ALWAYS=true
```

Usage
---
```bash
php artisan swaggerator:generate TagName requestName path/to/request
```

**Parameters**

`-p|pram` to add parameters
```bash
php artisan swaggerator:generate TagName requestName path/to/request -pmail --pram=password

```

`--method` to add parameters
```bash
php artisan swaggerator:generate TagName requestName path/to/request --method=Get

```



Output
---
output folder  `app/Document/TagName.php`
```php
/**
 * @OA\Post(
 *      path="/path/to/request ",
 *      operationId="---",
 *      tags={"TagName"},
 *      summary="----",
 *      description="---",
 *      @OA\Parameter(
 *          name="email",
 *          description="----",
 *          required=true,
 *          in="path",
 *          @OA\Schema(
 *              type="integer"
 *          )
 *      ),
 *      @OA\Parameter(
 *          name="password",
 *          description="----",
 *          required=true,
 *          in="path",
 *          @OA\Schema(
 *              type="integer"
 *          )
 *      ),
 *      @OA\Response(
 *          response=200,
 *          description="successful operation"
 *       ),
 *      @OA\Response(response=400, description="Bad request"),
 *      @OA\Response(response=404, description="Resource Not Found"),
 * )
 */
```

after running the command go to 
```bash
localost:8000/api/documentation
```


Changelog
---
Check [CHANGELOG](CHANGELOG.md) for the changelog

Testing
---
To run tests use

    $ composer test

Contributing
---


Security
---
If you discover any security related issues, please email p.pator@outlook.com or use the issue tracker of GitHub.

About
---

License
---
The MIT License (MIT). Please see [License File](LICENSE) for more information.

CHANGELOG.md
CHANGELOG
===

This changelog contains all notable change of the swaggerator package

tests/TestCase.php
<?php

namespace Peterzaccha\Swaggerator\Test;

use Peterzaccha\Swaggerator\SwaggeratorServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

abstract class TestCase extends Orchestra
{

    /**
     * Setup the test environment.
     */
    public function setUp()
    {
        parent::setUp();
    }

    /**
     * add the package provider
     *
     * @param $app
     * @return array
     */
    protected function getPackageProviders($app)
    {
        return [SwaggeratorServiceProvider::class];
    }

    /**
     * Define environment setup.
     *
     * @param  \Illuminate\Foundation\Application  $app
     * @return void
     */
    protected function getEnvironmentSetUp($app)
    {
        // Setup default database to use sqlite :memory:
        $app['config']->set('database.default', 'testing');
        $app['config']->set('database.connections.testing', [
            'driver'   => 'sqlite',
            'database' => ':memory:',
            'prefix'   => '',
        ]);
    }
}

src/Generate.php
<?php

namespace Peterzaccha\Swaggerator;

use Illuminate\Console\Command;

class Generate extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'swaggerator:generate {tag} {name} {url} {--method=Get} {--p|prams=*}';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Generate Swagger annotations';

    /**
     * Create a new command instance.
     *
     * @return void
     */
    public function __construct()
    {
        parent::__construct();
    }

    /**
     * Execute the console command.
     *
     * @return mixed
     */
    public function handle()
    {
        $tag = $this->argument('tag');
        $name = $this->argument('name');
        $dir = app_path().'/Document';
        $filePath = $dir.'/'.$tag.'.php';

        if (!file_exists($dir)) {
            mkdir($dir, 0777, true);
        }

        if (!file_exists($filePath)) {
             file_put_contents($filePath,'<?php
            ');
        }


        $schema = $this->schemaGenerator($tag,$name,$this->option('method'),$this->argument('url'),$this->option('prams'));

         file_put_contents($filePath,$schema,FILE_APPEND);
         $this->info($tag.' Created');

    }

    private function schemaGenerator($tag,$name,$method,$url,$prames){

        $default = [' *     @OA\Parameter(
 *          name="Content-Type",
 *          description="----",
 *          required=true,
 *          in="header",
 *          @OA\Schema(
 *              type="string",
 *           default="application/json",
 *          )','
 *      ),@OA\Parameter(
 *          name="X-Requested-With",
 *          description="----",
 *          required=true,
 *          in="header",
 *          @OA\Schema(
 *              type="string",
 *           default="XMLHttpRequest",
 *          )
 *      ),'];

        $prames =array_map(function ($pram){
            return '
                 *      @OA\Parameter(
                 *          name="'.$pram.'",
                 *          description="----",
                 *          required=true,
                 *          in="path",
                 *          @OA\Schema(
                 *              type="integer"
                 *          )
                 *      ),
            ';
        },$prames);

        $prames = array_merge($prames,$default);
        return '/**
                 * @OA\\'.$method.'(
                 *      path="/'.$url.'",
                 *      operationId="---",
                 *      tags={"'.$tag.'"},
                 *      summary="----",
                 *      description="---",
                 '.implode('',$prames).'
                 *      @OA\Response(
                 *          response=200,
                 *          description="successful operation"
                 *       ),
                 *      @OA\Response(response=400, description="Bad request"),
                 *      @OA\Response(response=404, description="Resource Not Found"),
                 *      security={
                 *         {
                 *             "oauth2_security_example": {"write:projects", "read:projects"}
                 *         }
                 *     },
                 * )
                 */
                ';


    }
}
